README.md
# ColorWithHSV-iOS

[![Version](https://img.shields.io/cocoapods/v/ColorWithHSV.svg?style=flat-square)](http://cocoapods.org/pods/ColorWithHSV)
[![License](https://img.shields.io/cocoapods/l/ColorWithHSV.svg?style=flat-square)](http://cocoapods.org/pods/ColorWithHSV)
[![Platform](https://img.shields.io/cocoapods/p/ColorWithHSV.svg?style=flat-square)](http://cocoapods.org/pods/ColorWithHSV)
[![CocoaPods](https://img.shields.io/cocoapods/metrics/doc-percent/ColorWithHSV.svg?style=flat-square)](http://cocoapods.org/pods/ColorWithHSV)
[![Build Status](https://img.shields.io/travis/GabrielMassana/ColorWithHSV-iOS/master.svg?style=flat-square)](https://travis-ci.org/GabrielMassana/ColorWithHSV-iOS)

##   What is it?

Swift Extension to convert a HSV colour values into a UIColor Objects.

HSV stands for hue, saturation, and value, and is also often called HSB (B for brightness). 

Formula from [Wikipedia](https://en.wikipedia.org/wiki/HSL_and_HSV#From_HSV).

## Installation

#### Podfile

```ruby
platform :ios, '8.0'
use_frameworks!

pod 'ColorWithHSV', '~> 0.0'
```

Then, run the following command:

```bash
$ pod install
```

#### Old school

Drag into your project the folder `/ColorWithHSV-iOS`. That's all.

## Example Swift

```swift
        let black: UIColor = UIColor.colorWithHSV(hue: 0.0, saturation: 0.0, value: 0.0)
        let white: UIColor = UIColor.colorWithHSV(hue: 0.0, saturation: 0.0, value: 1.0)
        let red: UIColor = UIColor.colorWithHSV(hue: 0.0, saturation: 1.0, value: 1.0)
        let green: UIColor = UIColor.colorWithHSV(hue: 120.0, saturation: 1.0, value: 1.0)
        let blue: UIColor = UIColor.colorWithHSV(hue: 240.0, saturation: 1.0, value: 1.0)
```
## Example Objective-C

```objc
#import "ColorWithHSV-Swift.h"

...
    UIColor *black = [UIColor hsv_colorWithHue:0.0 saturation:0.0 value:0.0];
    UIColor *white = [UIColor hsv_colorWithHue:0.0 saturation:0.0 value:1.0];
    UIColor *red = [UIColor hsv_colorWithHue:0.0 saturation:1.0 value:1.0];
    UIColor *green = [UIColor hsv_colorWithHue:120.0 saturation:1.0 value:1.0];
    UIColor *blue = [UIColor hsv_colorWithHue:240.0 saturation:1.0 value:1.0];
    
```
## License

ColorWithHSV-iOS is released under the MIT license. Please see the file called LICENSE.

## Versions

```bash
$ git tag -a 1.0.0 -m 'Version 1.0.0'

$ git push --tags
```

## Author

Gabriel Massana

##Found an issue?

Please open a [new Issue here](https://github.com/GabrielMassana/ColorWithHSV-iOS/issues/new) if you run into a problem specific to ColorWithHSV-iOS, have a feature request, or want to share a comment.


CHANGELOG.md
#### 1.0.2

 - First release
 - Extension will create a UIColor object based on given HSV values.
    - Hue value must be between 0 and 360.
    - Saturation value must be between 0 and 1.
    - Value value must be between 0 and 1.
 
 - Added some Tests




ColorWithHSV-iOS/UIColor+HSV.swift
//
//  UIColor+HSV.swift
//  ColorWithHSV
//
//  Created by GabrielMassana on 03/04/2016.
//  Copyright © 2016 GabrielMassana. All rights reserved.
//
//  Formula: https://en.wikipedia.org/wiki/HSL_and_HSV#From_HSL
//

import UIKit

public extension UIColor {
    
    //MARK: - Public method
    
    /**
    Creates UIColor object based on given HSV values.
    
    - parameter hue: CGFloat with the hue value. Hue value must be between 0 and 360.
    - parameter saturation: CGFloat with the saturation value. Saturation value must be between 0 and 1.
    - parameter value: CGFloat with the value value. Value value must be between 0 and 1.
    
    - returns: A UIColor from the given HSV values.
    */
    @objc(hsv_colorWithHue:saturation:value:)
    public class func colorWithHSV(hue hue: CGFloat, saturation: CGFloat, value: CGFloat) -> UIColor? {
        
        var r: CGFloat = 0.0
        var g: CGFloat = 0.0
        var b: CGFloat = 0.0
        
        guard hue <= 360.0 && hue >= 0.0 else { return nil }
        guard saturation <= 1.0 && saturation >= 0.0 else { return nil }
        guard value <= 1.0 && value >= 0.0 else { return nil }
        
        let chroma: CGFloat = value * saturation
        let h60: CGFloat = hue / 60.0
        let x: CGFloat = chroma * (1 - abs((h60 % 2) - 1))
        
        if (h60 < 1) {
            
            r = chroma
            g = x
        }
        else if (h60 < 2)
        {
            r = x
            g = chroma
        }
        else if (h60 < 3)
        {
            g = chroma
            b = x
        }
        else if (h60 < 4)
        {
            g = x
            b = chroma
        }
        else if (h60 < 5)
        {
            r = x
            b = chroma
        }
        else if (h60 < 6)
        {
            r = chroma
            b = x
        }
        
        let m: CGFloat = value - chroma
        
        r = r + m
        g = g + m
        b = b + m
        
        return UIColor(red: r, green: g, blue: b, alpha: 1.0)
    }
}

ColorWithHSV/Example/AppDelegate.swift
//
//  AppDelegate.swift
//  ColorWithHSV
//
//  Created by GabrielMassana on 03/04/2016.
//  Copyright © 2016 GabrielMassana. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(application: UIApplication, didFinishLaunchingWithOptions launchOptions: [NSObject: AnyObject]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(application: UIApplication) {
        // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



ColorWithHSVTests/ColorWithHSVTests.swift
//
//  ColorWithHSVTests.swift
//  ColorWithHSVTests
//
//  Created by GabrielMassana on 03/04/2016.
//  Copyright © 2016 GabrielMassana. All rights reserved.
//

import XCTest

@testable import ColorWithHSV

class colorWithHSVTests: XCTestCase {
    
    override func setUp() {
        
        super.setUp()
    }
    
    override func tearDown() {
        
        super.tearDown()
    }
    
    //MARK: - Valid
    
    func test_colorWithHSV_newObjectReturned_minimum() {
        
        let color = UIColor.colorWithHSV(hue: 0.0, saturation: 0.0, value: 0.0)
        
        XCTAssertNotNil(color, "A valid Color object wasn't created");
    }
    
    func test_colorWithHSV_newObjectReturned_maximum() {
        
        let color = UIColor.colorWithHSV(hue: 360.0, saturation: 1.0, value: 1.0)
        
        XCTAssertNotNil(color, "A valid Color object wasn't created");
    }
    
    //MARK: - NoValid
    
    func test_colorWithHSV_outOfRangeHue_over() {
        
        let color = UIColor.colorWithHSV(hue: 361.0, saturation: 0.5, value: 0.5)
        
        XCTAssertNil(color, "A valid Color object was created");
    }
    
    func test_colorWithHSV_outOfRangeHue_under() {
        
        let color = UIColor.colorWithHSV(hue: -1.0, saturation: 0.5, value: 0.5)
        
        XCTAssertNil(color, "A valid Color object was created");
    }
    
    func test_colorWithHSV_outOfRangeSaturation_over() {
        
        let color = UIColor.colorWithHSV(hue: 180.0, saturation: 1.1, value: 0.5)
        
        XCTAssertNil(color, "A valid Color object was created");
    }
    
    func test_colorWithHSV_outOfRangeSaturation_under() {
        
        let color = UIColor.colorWithHSV(hue: 180.0, saturation: -0.1, value: 0.5)
        
        XCTAssertNil(color, "A valid Color object was created");
    }
    
    func test_colorWithHSV_outOfRangevalue_over() {
        
        let color = UIColor.colorWithHSV(hue: 180.0, saturation: 0.5, value: 1.1)
        
        XCTAssertNil(color, "A valid Color object was created");
    }
    
    func test_colorWithHSV_outOfRangevalue_under() {
        
        let color = UIColor.colorWithHSV(hue: 180.0, saturation: 0.5, value: -0.1)
        
        XCTAssertNil(color, "A valid Color object was created");
    }
    
    //MARK: - SpecificColor
    
    func test_colorWithHSV_red() {
        
        let redColor = UIColor.colorWithHSV(hue: 0.0, saturation: 1.0, value: 1.0)
        
        XCTAssertEqual(redColor, UIColor.redColor(), "A red Color object wasn't created");
    }
    
    func test_colorWithHSV_green() {
        
        let greenColor = UIColor.colorWithHSV(hue: 120.0, saturation: 1.0, value: 1.0)
        
        XCTAssertEqual(greenColor, UIColor.greenColor(), "A green Color object wasn't created");
    }
    
    func test_colorWithHSV_blue() {
        
        let blueColor = UIColor.colorWithHSV(hue: 240.0, saturation: 1.0, value: 1.0)
        
        XCTAssertEqual(blueColor, UIColor.blueColor(), "A blue Color object wasn't created");
    }
}


ColorWithHSV/Example/ViewController.swift
//
//  ViewController.swift
//  ColorWithHSV
//
//  Created by GabrielMassana on 03/04/2016.
//  Copyright © 2016 GabrielMassana. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        
        super.viewDidLoad()
        
        let black: UIColor = UIColor.colorWithHSV(hue: 0, saturation: 0, value: 0)!
        let white: UIColor = UIColor.colorWithHSV(hue: 0, saturation: 0, value: 1)!
        let red: UIColor = UIColor.colorWithHSV(hue: 0, saturation: 1, value: 1)!
        let green: UIColor = UIColor.colorWithHSV(hue: 120, saturation: 1, value: 1)!
        let blue: UIColor = UIColor.colorWithHSV(hue: 240, saturation: 1, value: 1)!
        
        print("black", black)
        print("white", white)
        print("red", red)
        print("green", green)
        print("blue", blue)
    }
}

